## GPS Accuracy Using Root Mean Squared Error (RMSE)

In [8]:
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd
import math

data = pd.read_excel('gps_data.xlsx')
lat = data['Latitude']
lon = data['Longitude']
true_lat = data['True Latitude']
true_lon = data['True Longitude']
true_dist, true_dist1, true_dist2, true_dist3 = [], [], [], []
dist1, dist2, dist3 = [], [], []
mse1, mse2, mse3 = [], [], []
loc_rmse = []

dlat = np.deg2rad(lat - true_lat)
dlon = np.deg2rad(lon - true_lon)
lat = np.deg2rad(lat)
lon = np.deg2rad(lon)
true_lat = np.deg2rad(true_lat)
true_lon = np.deg2rad(true_lon)

for i in lat:
    true_dist_value = 0
    true_dist.append(true_dist_value)

# Haversine Formula
rEarth = 6371000
a = (np.sin(dlat / 2) ** 2) + np.cos(true_lat) * np.cos(lat) * ((np.sin(dlon / 2) / 2) ** 2)
c = 2 * np.arcsin(np.sqrt(a))
dist = rEarth * c

### First Location (Pililla, Rizal)

In [9]:
data[0:30]

Latitude   Longitude  True Latitude  True Longitude         Location
0   14.430120  121.337312      14.430289      121.337339  Pililla,  Rizal
1   14.430121  121.337312      14.430289      121.337339              NaN
2   14.430121  121.337309      14.430289      121.337339              NaN
3   14.430124  121.337305      14.430289      121.337339              NaN
4   14.430125  121.337304      14.430289      121.337339              NaN
5   14.430121  121.337299      14.430289      121.337339              NaN
6   14.430120  121.337293      14.430289      121.337339              NaN
7   14.430117  121.337290      14.430289      121.337339              NaN
8   14.430118  121.337291      14.430289      121.337339              NaN
9   14.430119  121.337291      14.430289      121.337339              NaN
10  14.430119  121.337289      14.430289      121.337339              NaN
11  14.430120  121.337289      14.430289      121.337339              NaN
12  14.430120  121.337291      14.430289      121.337339              NaN
13  14.430120  121.337292      14.430289      121.337339              NaN
14  14.430125  121.337295      14.430289      121.337339              NaN
15  14.430130  121.337297      14.430289      121.337339              NaN
16  14.430131  121.337296      14.430289      121.337339              NaN
17  14.430133  121.337298      14.430289      121.337339              NaN
18  14.430134  121.337301      14.430289      121.337339              NaN
19  14.430134  121.337304      14.430289      121.337339              NaN
20  14.430132  121.337302      14.430289      121.337339              NaN
21  14.430130  121.337301      14.430289      121.337339              NaN
22  14.430127  121.337300      14.430289      121.337339              NaN
23  14.430126  121.337299      14.430289      121.337339              NaN
24  14.430120  121.337299      14.430289      121.337339              NaN
25  14.430116  121.337299      14.430289      121.337339              NaN
26  14.430115  121.337304      14.430289      121.337339              NaN
27  14.430120  121.337310      14.430289      121.337339              NaN
28  14.430120  121.337313      14.430289      121.337339              NaN
29  14.430122  121.337316      14.430289      121.337339              NaN

### First Location RMSE

In [10]:
dist1 = dist[0:30]
true_dist1 = true_dist[0:30]
mse1 = mean_squared_error(dist1, true_dist1)
rmse = np.sqrt(mse1)
print(rmse)
loc_rmse.append(np.sqrt(mse1))

18.521240540130293


### Second Location (Pineda, Pasig City)

In [11]:
data[30:60]

Latitude   Longitude  True Latitude  True Longitude            Location
30  14.566052  121.058144       14.56603       121.05823  Pineda, Pasig City
31  14.566055  121.058136       14.56603       121.05823                 NaN
32  14.566056  121.058136       14.56603       121.05823                 NaN
33  14.566058  121.058128       14.56603       121.05823                 NaN
34  14.566060  121.058128       14.56603       121.05823                 NaN
35  14.566063  121.058121       14.56603       121.05823                 NaN
36  14.566065  121.058121       14.56603       121.05823                 NaN
37  14.566067  121.058121       14.56603       121.05823                 NaN
38  14.566068  121.058121       14.56603       121.05823                 NaN
39  14.566068  121.058121       14.56603       121.05823                 NaN
40  14.566065  121.058121       14.56603       121.05823                 NaN
41  14.566061  121.058128       14.56603       121.05823                 NaN
42  14.566059  121.058128       14.56603       121.05823                 NaN
43  14.566057  121.058128       14.56603       121.05823                 NaN
44  14.566057  121.058128       14.56603       121.05823                 NaN
45  14.566057  121.058128       14.56603       121.05823                 NaN
46  14.566055  121.058128       14.56603       121.05823                 NaN
47  14.566055  121.058121       14.56603       121.05823                 NaN
48  14.566059  121.058121       14.56603       121.05823                 NaN
49  14.566061  121.058121       14.56603       121.05823                 NaN
50  14.566063  121.058128       14.56603       121.05823                 NaN
51  14.566064  121.058128       14.56603       121.05823                 NaN
52  14.566068  121.058128       14.56603       121.05823                 NaN
53  14.566067  121.058128       14.56603       121.05823                 NaN
54  14.566067  121.058128       14.56603       121.05823                 NaN
55  14.566067  121.058128       14.56603       121.05823                 NaN
56  14.566066  121.058121       14.56603       121.05823                 NaN
57  14.566067  121.058113       14.56603       121.05823                 NaN
58  14.566069  121.058105       14.56603       121.05823                 NaN
59  14.566073  121.058098       14.56603       121.05823                 NaN

### Third Location RMSE

In [12]:
dist2 = dist[30:60]
true_dist2 = true_dist[30:60]
mse2 = mean_squared_error(dist2, true_dist2)
rmse = np.sqrt(mse2)
print(rmse)
loc_rmse.append(rmse)

6.791612817518172


### Third Location (Greenwoods, Pasig City)

In [14]:
data[60:90]

Latitude   Longitude  True Latitude  True Longitude  \
60  14.562879  121.098138      14.562817      121.098087   
61  14.562879  121.098139      14.562817      121.098087   
62  14.562880  121.098140      14.562817      121.098087   
63  14.562880  121.098140      14.562817      121.098087   
64  14.562881  121.098140      14.562817      121.098087   
65  14.562882  121.098141      14.562817      121.098087   
66  14.562882  121.098140      14.562817      121.098087   
67  14.562882  121.098140      14.562817      121.098087   
68  14.562883  121.098139      14.562817      121.098087   
69  14.562883  121.098139      14.562817      121.098087   
70  14.562885  121.098139      14.562817      121.098087   
71  14.562886  121.098140      14.562817      121.098087   
72  14.562886  121.098140      14.562817      121.098087   
73  14.562887  121.098140      14.562817      121.098087   
74  14.562888  121.098140      14.562817      121.098087   
75  14.562889  121.098140      14.562817      121.098087   
76  14.562890  121.098140      14.562817      121.098087   
77  14.562891  121.098141      14.562817      121.098087   
78  14.562893  121.098143      14.562817      121.098087   
79  14.562893  121.098144      14.562817      121.098087   
80  14.562894  121.098144      14.562817      121.098087   
81  14.562894  121.098145      14.562817      121.098087   
82  14.562895  121.098146      14.562817      121.098087   
83  14.562895  121.098148      14.562817      121.098087   
84  14.562895  121.098150      14.562817      121.098087   
85  14.562895  121.098151      14.562817      121.098087   
86  14.562895  121.098151      14.562817      121.098087   
87  14.562895  121.098151      14.562817      121.098087   
88  14.562895  121.098151      14.562817      121.098087   
89  14.562894  121.098151      14.562817      121.098087   

                   Location  
60  Greenwoods , Pasig City  
61                      NaN  
62                      NaN  
63                      NaN  
64                      NaN  
65                      NaN  
66                      NaN  
67                      NaN  
68                      NaN  
69                      NaN  
70                      NaN  
71                      NaN  
72                      NaN  
73                      NaN  
74                      NaN  
75                      NaN  
76                      NaN  
77                      NaN  
78                      NaN  
79                      NaN  
80                      NaN  
81                      NaN  
82                      NaN  
83                      NaN  
84                      NaN  
85                      NaN  
86                      NaN  
87                      NaN  
88                      NaN  
89                      NaN

In [15]:
dist3 = dist[60:90]
true_dist3 = true_dist[60:90]
mse3 = mean_squared_error(dist3, true_dist3)
rmse = np.sqrt(mse3)
print(rmse)
loc_rmse.append(np.sqrt(mse3))

8.50024212569389


## Remarks (not updated)

The RMSE serves as a guide to determine the accuracy of a GPS within a specific range. It tells the observer that their true location might be within *n* meters of the true location (*n* is the RMSE).

Based on the data gathered by the researchers, they have noticed that the RMSE calculated on the first location is **18.52 meters**, this suggests that the true value of the coordinate lies within that range. A factor that could have caused the low accuracy are the presence of trees in the area. On the other hand, the accuracy on the second location is within **6.79 meters**. This is a normal range for a low-cost commercial GPS, however, according to the datasheet the NEO-M8N has an accuracy within 2.5 meters (under optimal conditions). It might be good to note that this was tested in a residential area (more reading should be done to understand the reasons why GPS signals are weakened).